### ISSUE TITLE
5-6x slower performance between v2 and v3

### ISSUE THREAD

**Issue by ofirnk** [explosion/spaCy#9858]
## Regression
With Spacy 2 we've experienced 5-6 times faster processing of documents, with same model "size" (not transformer).

#### With `nlp(text)`
* In Spacy 2: `For text of length 789417 (~780KB) it takes ~56 seconds for full parse`
With pipeline = ['tagger', 'parser', 'ner']
* In Spacy 3: `For text of length 789417 (~780KB) it takes ~303 seconds for full parse`
With pipeline = ['tok2vec', 'tagger', 'parser', 'attribute_ruler', 'lemmatizer', 'ner']

#### With `nlp.pipe([chunks)]` and 16 chunks
* In Spacy 2: `For text of length 789417 (~780KB) it takes ~38 seconds for full parse`
* In Spacy 3: `For text of length 789417 (~780KB) it takes ~203 seconds for full parse`

## How to reproduce the behaviour

Created a Colab with dropdown to choose Spacy <3 or >3
(Please choose "restart runtime and run" after switching between spacy versions)
https://colab.research.google.com/drive/10XUpSQ6ir5-DttWJsUN5SjjVQfCDmglN?usp=sharing

I've plot the analyze time as relative to text length..

![image](https://user-images.githubusercontent.com/8449713/146670941-b43f0ff6-3415-4ae0-9d2c-12e823f09a39.png)
(ratio is attached to right axis, so it goes 2->3->4->5+)

```python
#@title Spacy 3 vs 2: benchmark large text

#@markdown **1. Download a rather large document**

import urllib
import urllib.request
large_text_file_url = "https://www.gutenberg.org/files/1342/1342-0.txt"

urllib.request.urlretrieve(large_text_file_url, "large.txt")  # download
large_txt = urllib.request.urlopen(large_text_file_url).read().decode('utf-8')

print(f"Length of large_txt={len(large_txt)}")
text_to_parse_dropdown = "large" #@param ["large", "small"]
text_to_parse = large_txt if text_to_parse_dropdown == "large" else "Hello"
print(f"Length of text_to_parse={len(text_to_parse)}")

#@markdown **2. Install spacy, download model, import & load model**
spacy_version = "spacy\u003C3" #@param ["spacy\u003C3", "spacy>3"]
import os; os.environ["spacy_version"] = spacy_version
!pip install "$spacy_version"
!python -m spacy download en_core_web_md
import en_core_web_md
nlp = en_core_web_md.load()

#@markdown **3. Run full parse, benchmark**
import timeit
results = timeit.timeit(lambda: nlp(text_to_parse), number=3)
print(results)
print(f'For text of length {len(text_to_parse)} it took {results}')
results
```
### Note: Disabling modules from Spacy 3 model doesn't seem to help
(using `disable = ['tok2vec', 'attribute_ruler', 'lemmatizer']` in colab)

## Your Environment

* Operating System: Ubuntu / WSL@Windows / Colab - all reproduce
* Python Version Used: 2 / 3
* spaCy Version Used: 2.3 / 3.2
* Environment Information: Cloudy

**Comment by honnibal** [comment]
Can you run your benchmark with `nlp.pipe()`? We normally benchmark throughput with batching, so it's possible that there's a performance regression on the unbatched case. But if you can use batching it will greatly improve your throughput.

**Comment by adrianeboyd** [comment]
Also, don't disable `tok2vec` since it's needed for `tagger` and `parser`, so if you time without it it's not realistic (you'll be timing a lot of nonsense annotation).

**Comment by ofirnk** [comment]
Thanks for the quick replies!

> Is it possible to presegment your texts so that you can run batched using `nlp.pipe()`? We normally benchmark throughput with batching, so it's possible that there's a performance regression on the unbatched case. But if you can use batching it will greatly improve your throughput.

I've tried it earlier and recorder around 2x degradation, but I'll include concrete numbers and share the code in colab.
It seems there's a big regression in speed in single threaded, so I'd be surprised if there's no major hit also on the batched version.
Will share in next 1-2 hours.

> Also, don't disable `tok2vec` since it's needed for `tagger` and `parser`, so if you time without it it's not realistic (you'll be timing a lot of nonsense annotation).

Times are without disabling any pipeline elements - so 300 seconds vs 56 is with "vanilla" model

**Comment by ofirnk** [comment]
> Is it possible to presegment your texts so that you can run batched using `nlp.pipe()`? We normally benchmark throughput with batching, so it's possible that there's a performance regression on the unbatched case. But if you can use batching it will greatly improve your throughput.

Updated the post with `nlp.pipe()` numbers - similar 5-6x performance hit (38 vs 203 seconds)

**Comment by adrianeboyd** [comment]
I used the speed benchmark here to do comparisons between v2.3 and v3.2 for `en_core_web_md`:

https://github.com/explosion/projects/tree/v3/benchmarks/speed

On shorter texts (the default in the benchmark) the speed difference for `en_core_web_md` is ~5%. I can replicate that on longer texts it's closer to 50%. My guess is that it's probably related to the transition parser (`parser` and/or `ner`), but we would have to do some more detailed profiling.

Generally there's little practical benefit to processing longer texts because little information beyond the paragraph level is useful for any of the pipeline components, but this kind of speed regression is definitely something we want to take a closer look at.

**Comment by ofirnk** [comment]
@adrianeboyd What size of text did you use when getting the 50% difference? 
(is it 50% as in 2x, or as 1.5x?)

**Comment by adrianeboyd** [comment]
I mean the speed was 50% of v2.3, so 2x. The short texts were 33 words/text on average (from the reddit sample downloaded as a project asset) and the long version was just quick a hack that repeated each short text 500x, so ~16k words/text.

**Comment by adrianeboyd** [comment]
There is some quadratic complexity in the transition parser, but the changes between v2.3 and v3.2 are not huge, so I wouldn't have expected this particular difference. It may be something else in `thinc`, for instance.

**Comment by ofirnk** [comment]
Ok it was important enough so I've plot the analyze time as relative to text length..

![image](https://user-images.githubusercontent.com/8449713/146372238-af04aaac-b058-49f7-8838-2b9294a3257e.png)

<details> 
  <summary>Data is here (click) </summary>

text length | analyze time v3 | analyze time v2 | ratio
-- | -- | -- | --
10000 | 0.8 | 0.7 | 1.1
20000 | 1.6 | 1.6 | 1.0
30000 | 2.4 | 1.9 | 1.3
40000 | 3.6 | 4.4 | 0.8
50000 | 4.7 | 3.1 | 1.5
60000 | 5.5 | 3.4 | 1.6
70000 | 7.3 | 4 | 1.8
80000 | 8.5 | 4.3 | 2.0
90000 | 9.7 | 5 | 1.9
100000 | 10.9 | 5.6 | 1.9
110000 | 12.3 | 6.3 | 2.0
120000 | 13.5 | 6.8 | 2.0
130000 | 14.8 | 7.6 | 1.9
140000 | 16.2 | 8 | 2.0
150000 | 17.8 | 8.2 | 2.2
160000 | 19.7 | 8.9 | 2.2
170000 | 20.8 | 9.2 | 2.3
180000 | 23.1 | 10.1 | 2.3
190000 | 24.4 | 10.5 | 2.3
200000 | 26.4 | 10.8 | 2.4
210000 | 27.8 | 11.5 | 2.4
220000 | 33.9 | 11.6 | 2.9
230000 | 32.9 | 12.7 | 2.6
240000 | 34.8 | 13.1 | 2.7
250000 | 36.8 | 13.3 | 2.8
260000 | 39.2 | 13.9 | 2.8
270000 | 42 | 14.3 | 2.9
280000 | 45.3 | 15 | 3.0
290000 | 48.5 | 15.6 | 3.1
300000 | 54.8 | 16.2 | 3.4
310000 | 52.6 | 16.7 | 3.1
320000 | 51.1 | 17 | 3.0
330000 | 53.5 | 17.7 | 3.0
340000 | 55.6 | 18.5 | 3.0
350000 | 57.7 | 18.4 | 3.1
360000 | 60.7 | 18.7 | 3.2
370000 | 63.6 | 20.2 | 3.1
380000 | 70.9 | 23.7 | 3.0
390000 | 70 | 20 | 3.5
400000 | 73.4 | 21.1 | 3.5
410000 | 76.5 | 21.7 | 3.5
420000 | 77.5 | 21.7 | 3.6
430000 | 80.3 | 22.1 | 3.6
440000 | 83.1 | 23.5 | 3.5
450000 | 86.7 | 22.8 | 3.8
460000 | 94.7 | 23.7 | 4.0
470000 | 93.3 | 24.6 | 3.8
480000 | 96 | 25 | 3.8
490000 | 98.6 | 25.2 | 3.9
500000 | 103 | 25.6 | 4.0
510000 | 106.9 | 26.4 | 4.0
520000 | 113.6 | 26.9 | 4.2
530000 | 114.1 | 27.3 | 4.2
540000 | 117.1 | 28.6 | 4.1
550000 | 121.7 | 28.6 | 4.3
560000 | 125.3 | 29.2 | 4.3
570000 | 128.9 | 29.4 | 4.4
580000 | 137.3 | 30.1 | 4.6
590000 | 136.3 | 30.6 | 4.5
600000 | 140 | 31 | 4.5
610000 | 145.4 | 31.6 | 4.6
620000 | 153.3 | 32.3 | 4.7
630000 | 153.3 | 32.6 | 4.7
640000 | 155.9 | 33.9 | 4.6
650000 | 160.6 | 34.1 | 4.7
660000 | 168.9 | 33.7 | 5.0
670000 | 170.2 | 35.2 | 4.8
680000 | 175.2 | 35.3 | 5.0
690000 | 178.7 | 35.4 | 5.0
700000 | 187.9 | 36.3 | 5.2

</details>

**Comment by adrianeboyd** [comment]
Thanks for the analysis, we are looking into it!

**Comment by ofirnk** [comment]
Another update - it happens also for TRF model over GPU:
![image](https://user-images.githubusercontent.com/8449713/146670941-b43f0ff6-3415-4ae0-9d2c-12e823f09a39.png)

The ratio between analysis time of trf+gpu/v2 increases over time
<details> 
  <summary>Data is here (click) </summary>

text length | analyze time v2 | analyze time trf-gpu | ratio trf-gpu/v2
-- | -- | -- | --
10000 | 0.7 | 2.1 | 3.0
20000 | 1.6 | 4.0 | 2.5
30000 | 1.9 | 6.2 | 3.3
40000 | 4.4 | 8.6 | 2.0
50000 | 3.1 | 10.9 | 3.5
60000 | 3.4 | 13.6 | 4.0
70000 | 4 | 16.1 | 4.0
80000 | 4.3 | 19.0 | 4.4
90000 | 5 | 22.2 | 4.4
100000 | 5.6 | 24.6 | 4.4
110000 | 6.3 | 28.3 | 4.5
120000 | 6.8 | 30.7 | 4.5
130000 | 7.6 | 34.1 | 4.5
140000 | 8 | 37.1 | 4.6
150000 | 8.2 | 41.0 | 5.0
160000 | 8.9 | 44.1 | 5.0
170000 | 9.2 | 47.8 | 5.2
180000 | 10.1 | 51.2 | 5.1
190000 | 10.5 | 54.3 | 5.2
200000 | 10.8 | 59.2 | 5.5
210000 | 11.5 | 62.9 | 5.5

</details>

**Comment by ofirnk** [comment]
> Another update - it happens also for TRF model over GPU: ![image](https://user-images.githubusercontent.com/8449713/146670941-b43f0ff6-3415-4ae0-9d2c-12e823f09a39.png)
> 
> The ratio between analysis time of trf+gpu/v2 increases over time
> Data is here (click)

Hey, any update about the regression?

**Comment by danieldk** [comment]
Sorry for the delay. We are looking into this, but the holidays have delayed this a bit, but it's definitely on our radar.

**Comment by danieldk** [comment]
Any chance (if it is not too much effort) you could test if #10048 solves the quadratic growth for you as well?

**Comment by ofirnk** [comment]
> Any chance (if it is not too much effort) you could test if #10048 solves the quadratic growth for you as well?

Sure, I'll try installing it on the colab I've shared earlier.
Is there some "assertion" I can do to verify I've installed and imported the correct branch?
Would spacy.version change to something other than 3.2.0?
`import spacy; print(spacy.__version__)`

**Comment by danieldk** [comment]
Fixed by #10048.
